src/main/java/hudson/tasks/JavadocArchiver.java

```diff
@@ -156,15 +156,35 @@ public String getIconFileName() {
          * Serves javadoc.
          */
         public void doDynamic(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException {
-            DirectoryBrowserSupport dbs = new DirectoryBrowserSupport(this, new FilePath(dir()), getTitle(), HELP_PNG, false);
-            if (new File(dir(), NOFRAMES_INDEX).exists() && Boolean.valueOf(
-                    System.getProperty(JavadocArchiver.class.getName() + ".useFramelessIndex", "true"))) {
-                /* If an overview-summary.html exists, serve that, unless the system property evaluates to false */
-                dbs.setIndexFileName(NOFRAMES_INDEX);
+            final File basedir = dir();
+            DirectoryBrowserSupport dbs = new DirectoryBrowserSupport(this, new FilePath(basedir), getTitle(), HELP_PNG, false);
+
+            String index = alternativeIndexFile(basedir);
+            if (index != null) {
+                dbs.setIndexFileName(index);
             }
+
             dbs.generateResponse(req, rsp, this);
         }
 
+        /**
+         * Default index.html can not be served by Jenkins safely both because Java 8 and older contains frames and Java 9 and newer uses Javascript.
+         *
+         * Suggest alternative file to start browsing from to overcome this limitation. See https://issues.jenkins-ci.org/browse/JENKINS-32619
+         */
+        private @CheckForNull String alternativeIndexFile(File basedir) {
+            // Suggest unless user insisted otherwise
+            boolean useFrameless = Boolean.parseBoolean(System.getProperty(JavadocArchiver.class.getName() + ".useFramelessIndex", "true"));
```

Comment:
I guess this could become a `@DataBoundSetter` if we feel the need, though on the other hand I would encourage users to just use `htmlpublisher` if they want anything other than defaults.

Reply:
Would it overcome the problems with frames or JS?

Reply:
No.